### PR TITLE
Moving all default settings to `.settings` file

### LINKED ISSUE #3555: Set up a .settings file
ref. discussion on #3534 

Let's put all the default settings in a .net-standard `.settings` file (i.e. accessible from project properties), at "application" level; that file would end up in the RD install folder and in no way would interfere with user settings. The contents of that file would be version-specific, and the file itself would be essentially treated as read-only: nothing ever writes to it.

This will require changes to all inspection classes (they all pass a hard-coded default severity through the base class constructor!), and remote a TON of shitty default-config setup code, notably for hotkeys.

Then the mechanics for loading user settings become brain-dead simple:

 - Load all the defaults (inspection severities, todo markers, hotkeys, indenter settings, etc.) from the application settings
 - See if there's a `rubberduck.config` user settings file (i.e. the xml file we're currently using); if so, load what can be loaded & override the loaded defaults. If user settings contain unknown or otherwise unloadable nodes, ignore them & keep processing.
 - Nothing else needs to change: user configurations keep updating the `rubberduck.config` xml file.

### PR DESCRIPTION
**Pull request by rkapka**
Closes #3555

### CODE REVIEW

**Comment by retailcoder** [comment]
Just to be clear: this PR is moving all *default settings* values to the .settings file, correct?

**Comment by rkapka** [comment]
@retailcoder Yes 😀. I will edit the title.

**Comment by retailcoder** [comment]
AppVeyor is claiming a conflict with `RubberduckTests/RubberduckTests.csproj`, not sure what's up with that. Otherwise LGTM

**Comment by rkapka** [comment]
Yeah, I didn't take a closer look that an inspection was added. The default severity is now inside the `.settings` file, so there is no constructor in `ParseTreeInspectionBase` taking a severity.

**Comment by rkapka** [comment]
I just noticed that I also moved the `InspectionType` into the `.settings` file, so at the moment it's specified there and in a property. Maybe the property could be defaulted to some value in the base class and be read from only if the inspection isn't present in the `.settings` file...

**Comment by retailcoder** [comment]
@rkapka but an inspection's `InspectionType` isn't a *setting*.. hmm, do we want them there though? Having inspection types in the settings makes it easier to visualize the groupings and distribution... I'm all for it! 👍

File: RetailCoder.VBE/Settings/WindowSettings.cs [hunk 1]
```diff
@@ -63,28 +65,22 @@ public bool IsWindowVisible(DockableToolwindowPresenter candidate)
...
-            {
-                return TestExplorerVisibleOnStartup;
-            }
-            if (candidate is ToDoExplorerDockablePresenter)
-            {
-                return TodoExplorerVisibleOnStartup;
-            }
-            //Oh. Hello. I have no clue who you are...
-            return false;
         }
```
Comment (retailcoder) [inline RetailCoder.VBE/Settings/WindowSettings.cs]: 👍

File: RetailCoder.VBE/Settings/CodeInspectionConfigProvider.cs [hunk 2]
```diff
@@ -1,41 +1,43 @@
...
+        private readonly CodeInspectionSettings _defaultSettings;
 
         public CodeInspectionConfigProvider(IPersistanceService<CodeInspectionSettings> persister, IEnumerable<IInspection> foundInspections)
         {
             _persister = persister;
-            _foundInspections = foundInspections;
+            _defaultSettings = new DefaultSettings<CodeInspectionSettings>().Default;
+
+            var nonDefaultInspections = foundInspections
+                .Where(inspection => !_defaultSettings.CodeInspections.Select(x => x.Name).Contains(inspection.Name));
```
Comment (retailcoder) [inline RetailCoder.VBE/Settings/CodeInspectionConfigProvider.cs]: I think that's iterating the inspections twice. `.Any()` could help here I think:

```csharp
var nonDefaultInspections = foundInspections
    .Where(inspection => !_defaultSettings.CodeInspections.Any(x => x.Name == inspection.Name));
```

File: RetailCoder.VBE/Settings/CodeInspectionSettings.cs [hunk 3]
```diff
@@ -126,42 +128,42 @@ public string SeverityLabel
...
-                   inspectionSetting.Severity == Severity;
+                   inspectionSetting.Name == Name;
         }
 
         public override int GetHashCode()
         {
             unchecked
             {
                 var hashCode = Name?.GetHashCode() ?? 0;
-                hashCode = (hashCode * 397) ^ (int)Severity;
```
Comment (retailcoder) [inline RetailCoder.VBE/Settings/CodeInspectionSettings.cs]: Holy crap, nice catch!! 😲

**Comment by Vogel612** [comment]
While InspectionTypes should not be a user-configurable setting, I think it's not a generally bad decision to make such settings configurable through our app settings. 

This should also remove the tests on all the inspections that check for a certain `InspectionType`... 
OTOH, this means that any plugged in inspections need to be configurable through this general app-configuration, which is not quite there yet.  
Then again we're not really ready for any plugins quite yet...

AFAIK we do not overwrite the Inspection's InspectionType with values read from settings, so in itself it's not used and should be removed...

**Comment by retailcoder** [comment]
@Vogel612  the `InspectionType` isn't read from settings, but the *inspection settings* UI **does** use that property for grouping the inspections, so it can't just be removed from the model.

Ideally the config file would drive what the `InspectionType` is for each inspection, but then individual inspections would have some way of specifying an `InspectionType` if we don't have a default for it... but then, this makes a rather confusing API: when implementing a new inspection, you immediately pick up that you need to provide an `InspectionType`, but not that you need to have it in the .settings file.... catch-22?

File: RetailCoder.VBE/Settings/CodeInspectionConfigProvider.cs (hunk 2 above)
Comment (rkapka) [reply RetailCoder.VBE/Settings/CodeInspectionConfigProvider.cs]: `Any` also iterates, so I extracted the names to a HashSet to avoid this.

Comment (retailcoder) [reply RetailCoder.VBE/Settings/CodeInspectionConfigProvider.cs]: It does, but `Select` is first projecting all elements into an `IEnumerable<string>`, which `Contains` then iterates. `Any` skips the projection part; HashSet seems overkill.

**Comment by rkapka** [comment]
@retailcoder How about making `InspectionType` a non-abstract, get/set property and defaulting it to `QualityIssues` or whichever value is most reasonable? `CodeInspectionConfigProvider` will set the property in its constructor using default settings. That way it would behave similarly to `Severity`.

**Comment by retailcoder** [comment]
@rkapka that can work... We just don't want the inspection types to end up in the serialized xml config file.

**Comment by Vogel612** [comment]
This will need to incorporate the new settings for #3717

**Comment by rkapka** [comment]
That would be all except leaving the inspection type setting out of the user-specific settings file. Excluding the property from serialization will prevent the `.settings` file from including the value in the default settings. I don't know how to solve this at the moment... It is possible to write a custom type converter and use `SettingsSerializeAs.String` to store a string representation of the setting in the `.settings` file, including the inspection type. But I am not sure how to embed such a string inside the XML setting for `CodeInspectionSettings` and if it is possible to serialize individual inspections as strings in the `.settings` file and as XML in the user-specific file.

For now the inspection type is visible in the user file, but it is ignored by the configuration provider. Any comments/ideas?

File: RetailCoder.VBE/Settings/CodeInspectionConfigProvider.cs (hunk 2 above)
Comment (rkapka) [reply RetailCoder.VBE/Settings/CodeInspectionConfigProvider.cs]: This part got removed.

**Comment by Vogel612** [comment]
Also fixes #2240

**Comment by retailcoder** [comment]
> For now the inspection type is visible in the user file, but it is ignored by the configuration provider. Any comments/ideas?

I think that's more than *good enough*... the configuration file is *intended* to be accessed from the settings dialog.. I guess we can consider it by-design.

**Pull request merged**